File: JavaAlgorithm/src/main/java/algo/tzashinorpu/SpecialSubject/BinaryTree/goodNodes_1448.java
package algo.tzashinorpu.SpecialSubject.BinaryTree;

import algo.tzashinorpu.SpecialSubject.TreeNode;

public class goodNodes_1448 {
    private int count = 0;

    public int goodNodes(TreeNode root) {
        helper(root, root.val);
        return count;
    }

    private void helper(TreeNode root, int val) {
        if (root == null) return;
        if (root.val >= val) count++;
        helper(root.left, Math.max(root.val, val));
        helper(root.right, Math.max(root.val, val));
    }
}

File: Clion/ZeroTrac/Tree/goodNodes_1448.cpp
//
// Created by TzashiNorpu on 10/28/2022.
//
#include "tree.h"

using namespace Tree;

int Solution::goodNodes(TreeNode *root) {
  int res = 0;
  function<void(TreeNode *, int)> helper = [&](TreeNode *node, int path_max) -> void {
      if (!node) return;
      if (path_max <= node->val) res++;
      path_max = max(path_max, node->val);
      helper(node->left, path_max);
      helper(node->right, path_max);
  };
  helper(root, root->val);
  return res;
}